sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/DoNotCallMethodsBase.cs

```diff
@@ -48,32 +48,39 @@ private void AnalyzeInvocation(SyntaxNodeAnalysisContext analysisContext)
         {
             var invocation = (InvocationExpressionSyntax)analysisContext.Node;
 
+            if (!ShouldReportPreCheck(invocation, analysisContext.SemanticModel))
+            {
+                return;
+            }
+
             var identifier = GetMethodCallIdentifier(invocation);
             if (identifier == null)
             {
                 return;
             }
 
-            var methodSignature = CheckedMethods.FirstOrDefault(method => method.Name.Equals(identifier.Value.ValueText));
-            if (methodSignature == null)
+            var methodCallSymbol = analysisContext.SemanticModel.GetSymbolInfo(identifier.Value.Parent).Symbol;
+            if (methodCallSymbol == null)
             {
                 return;
             }
 
-            var methodCallSymbol = analysisContext.SemanticModel.GetSymbolInfo(identifier.Value.Parent);
-            if (methodCallSymbol.Symbol == null ||
-                !methodCallSymbol.Symbol.ContainingType.ConstructedFrom.Is(methodSignature.ContainingType))
+            var disallowedMethodSignature = FindDisallowedMethodSignature(identifier.Value, methodCallSymbol);
+            if (disallowedMethodSignature == null)
             {
                 return;
             }
 
             if (ShouldReportOnMethodCall(invocation, analysisContext.SemanticModel))
             {
                 analysisContext.ReportDiagnosticWhenActive(Diagnostic.Create(Rule, identifier.Value.GetLocation(),
-                    methodSignature.ToShortName()));
+                    disallowedMethodSignature.ToShortName()));
             }
         }
 
+        protected virtual bool ShouldReportPreCheck(InvocationExpressionSyntax invocationSyntax,
```

Comment:
I would rename this method to something that express better the fact this is done before checking the method call (e.g. `IsInValidContext()`).